### ISSUE TITLE
Unify code of source information access for DWARF, PDB, dSYM

### ISSUE THREAD

**Issue by XVilka** [rizinorg/rizin#907]
Currently the code for handling source-level information access with DWARF, PDB, dSYM, etc is separate in every cases.
We should think about better abstract API for all of these. It will also affect config variables like `asm.dwarf`, etc.

See also:
- https://github.com/rizinorg/rizin/pull/880
- https://github.com/rizinorg/rizin/issues/653
- https://github.com/rizinorg/rizin/issues/3535

Good source of inspiration is the `DebugInfo` APIs from Binary Ninja: https://api.binary.ninja/binaryninja.debuginfo-module.html

**Comment by thestr4ng3r** [comment]
After #880, the line info is available in a format-agnistic way in `RzBinObject.lines` so from the perspective of querying the info after it has been loaded, this is already done.
Regarding loading itself, bin plugins can supply their line info using the `lines` callback. This is already nice, but it only really works when the debug info is bound to the binary format itself, like dex.
Dwarf however can occur in pretty much any format including elf and pe, so it doesn't work there. I could imagine additional debug info plugins that, if the bin plugin doesn't return lines, are tried one after another. Then dwarf would be one such plugin. (There were dbginfo plugins before, but they were not doing that)

**Comment by ret2libc** [comment]
> Dwarf however can occur in pretty much any format including elf and pe, so it doesn't work there.

As well as separate files completely. Not sure if it changes anything, but just keep it in mind.

**Comment by XVilka** [comment]
Even more, it can be served by a daemon: 

> With GNU Binutils 2.34 comes debuginfod support, which is the HTTP server catching our eye while the debuginfod server is distributed as part of the latest elfutils package. This isn't for a general purpose web server thankfully but is an HTTP server for distributing ELF/DWARF debugging information and source code. With debuginfod enabled, Binutils' `readelf` and `objdump` utilities can query the HTTP server(s) for debug files that cannot otherwise be found. Enabling this option requires building Binutils using `--with-debuginfod`. 

See https://developers.redhat.com/blog/2019/10/14/introducing-debuginfod-the-elfutils-debuginfo-server/ for more information.

- https://sourceware.org/elfutils/Debuginfod.html
- https://wiki.debian.org/Debuginfod

**Comment by XVilka** [comment]
@imbillow note, that also it would require some changes in the config variables - e.g. `asm.dwarf.` etc.
We should think what make sense to be generic, e.g. `asm.dbginfo`, etc and use these instead.
Just like we have `bin.dbginfo` already.

**Comment by XVilka** [comment]
TODO: Check also the Rust 1.70 new feature (e.g. `-Cdebuginfo=line-tables-only`), how it works with Rizin: https://github.com/rust-lang/rust/pull/109808/

**Comment by XVilka** [comment]
@imbillow turned out OMF also could have source code information available, currently it's not parsed:
- `librz/bin/formats/omf/*` - see `OMF_LINNUM` and `OMF_LINNUM32` for example. We should also add a basic support for those, and integrate it into this new source information access subsystem as well.
- [OMF: Relocatable Object Module Format Specification](https://github.com/rizinorg/rizin/files/11768394/Omfg.pdf)

Note, also we don't seem to parse types information available from OMF as well, probably could be done, similarly to DWARF and PDB.

**Comment by XVilka** [comment]
As most of this was complete, I am closing this issue, thanks @imbillow for the hard work! If something is missing, lets open a new issue.
